### PR TITLE
Lower shuffle to single-result form if possible.

### PR DESCRIPTION
**Pull request by jreiffers**
We currently always lower shuffle to the struct-returning variant. I saw some cases where this survived all the way through ptx, resulting in increased register usage. The easiest fix is to simply lower to the single-result version when the predicate is unused.

### CODE REVIEW

File: mlir/lib/Conversion/GPUToNVVM/LowerGpuOpsToNVVMOps.cpp [hunk 1]
```diff
@@ -176,14 +174,26 @@ struct GPUShuffleOpLowering : public ConvertOpToLLVMPattern<gpu::ShuffleOp> {
...
-    Value isActiveSrcLane = rewriter.create<LLVM::ExtractValueOp>(loc, shfl, 1);
-
-    rewriter.replaceOp(op, {shflValue, isActiveSrcLane});
+    if (predIsUsed) {
+      Value shflValue = rewriter.create<LLVM::ExtractValueOp>(loc, shfl, 0);
+      Value isActiveSrcLane =
+          rewriter.create<LLVM::ExtractValueOp>(loc, shfl, 1);
+      rewriter.replaceOp(op, {shflValue, isActiveSrcLane});
+    } else {
+      Value falseCst = rewriter.create<LLVM::ConstantOp>(loc, predTy, 0);
```
Comment (chsigg) [inline mlir/lib/Conversion/GPUToNVVM/LowerGpuOpsToNVVMOps.cpp:194]: Would it be better to not create a predicate value at all?
Instead of `rewriter.replaceOp()`, you could do `rewrite.replaceAllUsesWith(op.getResult(0), shfl)` plus `rewriter.eraseOp(op)`.

**Review by chsigg — COMMENTED** [review]
Looks good to me, I just have one minor nit.

**Review by chsigg — APPROVED** [review]

File: mlir/lib/Conversion/GPUToNVVM/LowerGpuOpsToNVVMOps.cpp (hunk 1 above)
Comment (jreiffers) [reply mlir/lib/Conversion/GPUToNVVM/LowerGpuOpsToNVVMOps.cpp]: Thanks, done.

Comment (joker-eph) [reply mlir/lib/Conversion/GPUToNVVM/LowerGpuOpsToNVVMOps.cpp:194]: Which cast is failing?

Comment (jreiffers) [reply mlir/lib/Conversion/GPUToNVVM/LowerGpuOpsToNVVMOps.cpp:194]: Unfortunately not, but the buildkite failed: https://buildkite.com/llvm-project/github-pull-requests/builds/49255#018e5fda-5040-4bf6-9d4d-b976d9799d20

Haven't yet been able to reproduce it.

Comment (jreiffers) [reply mlir/lib/Conversion/GPUToNVVM/LowerGpuOpsToNVVMOps.cpp:194]: Maybe I just messed up applying the fix? Let me try again.

Comment (joker-eph) [reply mlir/lib/Conversion/GPUToNVVM/LowerGpuOpsToNVVMOps.cpp]: That reproduced for me FYI:
```
# | #12 0x00005640f7453cc8 decltype(auto) llvm::cast<mlir::LLVM::LLVMArrayType, mlir::Type>(mlir::Type&) /home/mamini/projects/llvm-project2/llvm/include/llvm/Support/Casting.h:573:37
# | #13 0x00005640f71e5735 getInsertExtractValueElementType(mlir::Type, llvm::ArrayRef<long>) /home/mamini/projects/llvm-project2/mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1590:18
# | #14 0x00005640f71e55ca mlir::LLVM::ExtractValueOp::build(mlir::OpBuilder&, mlir::OperationState&, mlir::Value, llvm::ArrayRef<long>) /home/mamini/projects/llvm-project2/mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1645:9
# | #15 0x00005640f99a78e9 mlir::LLVM::ExtractValueOp mlir::OpBuilder::create<mlir::LLVM::ExtractValueOp, mlir::Value&, int>(mlir::Location, mlir::Value&, int&&) /home/mamini/projects/llvm-project2/mlir/include/mlir/IR/Builders.h:511:5
# | #16 0x00005640f9be1283 (anonymous namespace)::GPUShuffleOpLowering::matchAndRewrite(mlir::gpu::ShuffleOp, mlir::gpu::ShuffleOpAdaptor, mlir::ConversionPatternRewriter&) const /home/mamini/projects/llvm-project2/mlir/lib/Conversion/GPUToNVVM/LowerGpuOpsToNVVMOps.cpp:191:34
```

Comment (joker-eph) [reply mlir/lib/Conversion/GPUToNVVM/LowerGpuOpsToNVVMOps.cpp]: Actually that was on the previous commit that failed buildkite, the latest push seems fine.

Comment (jreiffers) [reply mlir/lib/Conversion/GPUToNVVM/LowerGpuOpsToNVVMOps.cpp]: Yep, looks like it was just me. Thanks for checking!
